### ISSUE TITLE
[Discussion] Reintroduce write-only signals

### ISSUE THREAD

**Issue by andriichernenko** [iZettle/Flow#73]
I am well aware of the fact that `WriteSignal` existed in Flow before, but was removed for the following reason:
> WriteSignal (previously called Sink) has been used, almost exclusively, as a setter closure, where it has been passed as an argument to Signal.bindTo(). In practice, only its setter has been used and never have we used the fact that a WriteSignal is also a signal.

However, when working on an app built with Flow according to MVVM architecture, I realized that it would be really nice to bring it back. I found it difficult to illustrate this with a code example, so I will try arguing theoretically.

---
View models are the centerpiece of MVVM. They accept inputs, process them, and produce outputs. Consumers should not care about the way processing is done, so view models can be defined by their inputs and outputs.

MVVM requires that view models have no knowledge of views (to make testing easier, among other things), relying on a binding mechanism that takes care of connecting view outputs and inputs to to the inputs and outputs of a view model instead. Therefore, having an easy to understand and use (i.e. ergonomical) binding mechanism is crucial.

When using Flow for binding, things are fairly straightforward with view model outputs. The outputs of a view model are `Signal<T>` and `ReadSignal<T>`, depending on whether an output is stateful or not. As for view model inputs, the following mechanisms can be used (proper disposal of subscriptions is omitted for the sake of brevity):
- functions (called later directly by view):
```swift
struct ViewModel {
    func doSomething(parameter: Int) {
        // processing
    }
}
```
- `ReadWriteSignal<T>`:
```swift
struct ViewModel {
    let doSomething: ReadWriteSignal<T>
    init() {
        doSomething.onValue { /* processing */ }
    }
}
```
- `Callbacker<T>`:
```swift
struct ViewModel {
    let doSomething: Callbacker<T>
    init() {
        doSomething.addCallback { /* processing */ }
    }
}
```

These mechanisms cover all possible situations. After all, Flow has been used for years, so I'd be surprised if there were unhandled cases. This made it possible to find a convincing code example. Nonetheless, each of the mechanisms but has its problems: 

#### Functions
- less convenient binding, must be called in `.onValue {  }` blocks, which is more verbose (`.onValue { viewModel.doSomething($0) }`) than just using `.bindTo(viewModel.doSomething)`;
- cannot be (idiomatically) composed with other signals in the view model;
- defined as methods on the view model, which may require keeping double definitions for the involved output signals (one private writable and one public read-only).

#### `ReadWriteSignal<T>`
- has state, which makes no sense in many cases (for example, input events);
- difficult to determine whether a `ReadWriteSignal<T>` defined on a view model is its input or output without inspecting its usage.

#### `Callbacker<T>`
This is actually the closest thing Flow has to what I am looking for, but it still is not perfect:
- less convenient binding, must be called in `.onValue { }` blocks, which is more verbose (`.onValue { viewModel.doSomething.callAll($0) }`) than just using `bindTo(viewModel.doSomething)`;
- cannot be composed with other signals in the view model without taking extra steps (wrapping it in a `Signal<T>`);
- it's name makes it stick out, although conceptually this is just a write-only signal.
- if we were to add a kind of `.writeOnly()` method to `ReadWriteSignal<T>` as a way to restrict the read access, it would be strange if it returned a `Callbacker<T>` instead of a `WriteSignal<T>`. Using 

During the internal discussion, @niil-ohlin suggested making `Callbacker<T>` a signal provider, which would invalidate the first 2 points, but the remaining 2 are still valid. I would argue that having `WriteSignal<T>` instead of `Callbacker<T>` would make the API of Flow cleaner, more symmetrical and easier to understand (I'm saying "instead", since there's probably no point in keeping both).

### Other libraries
I looked at RxSwift and the way it handles mutability. There is a concept of [`Subject`](http://reactivex.io/documentation/subject.html), which is
> a sort of bridge or proxy [...] that acts both as an observer and as an `Observable`. Because it is an observer, it can subscribe to one or more `Observable`s, and because it is an `Observable`, it can pass through the items it observes by reemitting them, and it can also emit new items. 

This looks like an equivalent of `ReadWriteSignal<T>` in Flow.

RxSwift supports giving write-only access to a `Subject` by transforming it into an [`AnyObserver<T>`](https://github.com/ReactiveX/RxSwift/blob/master/RxSwift/AnyObserver.swift). I looked at a few code examples, and there seems to be a pattern where inputs of a view model are defined using a combination of a private `Subject<T>` and a public `AnyObserver<T>` (examples: [\[1\]](https://medium.com/blablacar-tech/rxswift-mvvm-66827b8b3f10), [\[2\]](https://hackernoon.com/how-to-use-mvvm-coordinators-and-rxswift-7364370b7b95), [\[3\]](https://gist.github.com/IanKeen/04393f2f29d9c3e67025d96589fdfb97), and so on), so the idea of using write-only signals as view model inputs is not meritless.

----
To summarize, Flow is robust, but in some cases the APIs it provides are not optimal. One of these is the case of defining the input signals for view models. Reintroducing `WriteSignal<T>` should make this task easier. Admittedly, the gain in each particular case is small, but, due to pervasive use of bindings in MVVM, the benefits will add up.

Comments? Thoughts? Suggestions?

**Comment by mansbernhardt** [comment]
Hi Andrii, 

This was also shortly discussed in https://github.com/iZettle/Flow/pull/40

For me the confusion part of a `WriteSignal` when used as a `sink` is why the user besides providing/sinking values into it would need it to be a `Signal` and listen on value changes. Would the user like to know when itself is setting the values? Or do the user think someone else will change the value, and need to be aware of this? As the user don't know, he/she would probably expect the latter.

```swift
struct ViewModel {
   let valueDidChange: WriteSignal<Value>
}

// As a user I only need to notify when a value changes not to know when I myself update it (I know that already). Or should I expect the signal to report changes that I myself did not introduced?
model.valueDidChange.setValue(newValue) // Called directly
bag += control.bindTo(model.valueDidChange) // Called indirectly 

bag += valueDidChange.onValue { ... }. // Do I expect someone else to signal changes?
```

Where as, just using a function for reporting updates makes it clear that its only used to report changes and not for listen on those (or other) changes.

```swift
struct ViewModel {
   let valueDidChange: (Value) -> ()
}

model.valueDidChange(newValue) // Called directly
bag += control.bindTo(model.valueDidChange) // Called indirectly 
```

And considering the syntax, bindTo should work as expected as seen above as one of its overloads expects a `(Value) -> ()` function.

I think I need to see a convincing code example to better understand why you would need something else than a function to call to report updates?

**Comment by andriichernenko** [comment]
Sorry for taking so long to reply, I was on sick leave.

@jdmoreira `WriteSignal` will definitely need some kind of setter method, something like `setValue(aValue)` or `postValue(aValue)`. In this case, we will also need a getter method to make sure the API is symmetrical.

@mansbernhardt My intention is to subscribe to the input signals _inside_ the view model, as a way to derive its output signals from its input signals by applying signal operators:

```swift
struct ViewModel {
    let inputSignal: WriteSignal<...>
    let outputSignal: ReadSignal<...>
    private let internalSignal: ReadSignal<...>
}

extension ViewModel {
    init() {
        outputSignal = inputSignal
            .throttle(0.5)
            .withLatestFrom(internalSignal)
            .map { (_, value2) in ... }
    }
}
```

Of course, if I were to replace `inputSignal` with a method, I could still read the value of `internalSignal`, but throttling would be much less convenient (unless, of course, there is some smart, concise, and self-contained way to throttle calls to methods).

**Comment by CarlEkman** [comment]
I think one use case where `WriteSignal`s could be applied is logging/tracking. By making presentable flows conform to a protocol containing a `WriteSignal<Event>` we could easily dump user events from within presentation flows - without initial values or states.

The same sentiment could be achieved by a `Callbacker<Event>` or even an exposed function, but I think the APIs of our signals could be useful when observing the events from the outside. I'd also argue it's a pretty testable, and non-intrusive approach for logging, without needs for singletons or network logic within the flow.

I put down some of my thoughts in this gist, let me know your thoughts: https://gist.github.com/CarlEkman/3ed333b8a397791f507f1041edddb6b5

**Comment by mansbernhardt** [comment]
@andriichernenko What I don't like about exposing `inputSignal` as a `WriteSignal` is that the "signal" part of it is an implementation detail that we are leaking. The signal part just happens to be need by this particular implementation of `init`, but that might not always be true, and should not be part of the view models public interface. 

I'm also not sure who is the "owner" of `internalSignal`. By being private, from what part are we hiding it from. If you look at how view models are typically used in [Presentation](https://github.com/izettle/presentation), we have two users of the view model, the part the materialize the model (actual creates a view controller etc), and the parts that instantiates it (which actually could be several, such as for testing/mocking etc). The latter `init`'s are often location elsewhere and hence would not have access to privates as in your example.

 So going back to your example but using a function for inputSignal instead:

```swift
struct ViewModel {
    let inputSignal: (Value) -> Void
    let outputSignal: ReadSignal<...>
    private let internalSignal: ReadSignal<...>
}

extension ViewModel {
    init() {
        let inputCallback = Callbacker<Value>()
        inputSignal = inputCallback.callAll
        outputSignal = Signal(callbacker: inputCallback)
            .throttle(0.5)
            .withLatestFrom(internalSignal)
            .map { (_, value2) in ... }
    }
}
```

And has been suggested elsewhere, we could conform `Callbacker` to `SignalProvider` hence simplify it a bit:

```swift
extension ViewModel {
    init() {
        let inputCallback = Callbacker<Value>()
        inputSignal = inputCallback.callAll
        outputSignal = inputCallback
            .throttle(0.5)
            .withLatestFrom(internalSignal)
            .map { (_, value2) in ... }
    }
}
```

**Comment by mansbernhardt** [comment]
@CarlEkman Thanks for another example. I'm still trying to get my head on why `WriteSignal` would be a good idea. I have a feeling it might, but I just can't see it yet. So it's great to see more examples.

As for your particular example and it's use in combination with Presentation (iZettle style). I'm not sure why you would not instead inject or set up the logger in the `init`. Especially when you have several models in a hierarchy, it's not obvious who will setup the observing of your log signals. Wouldn't the most natural place to set it up be from the parent model's `init` and not from outside as in your example?  

Most concern I have about @andriichernenko example also applies to yours. I think we should strive for a view model to expose everything (but not more than) needed to materialize/present the model into a e.g. view controller. Having the logging part as a signal is leaking implementation details, similar to @andriichernenko example. And as your example is suggesting, the more convenient use would be your `func track()` method instead of working with the signal, that is, we are back to using a plain function.

**Comment by CarlEkman** [comment]
Fair point @mansbernhardt :) I suppose I just fancy the idea of treating events like a stream of some kind. I also have a feeling `WriteSignal` may be useful, but I agree that we're pretty much grasping at straws here.

Honestly it's mainly the naming scheme of the signals that bother me a bit, as they imply that a write-only signal is missing. If `ReadSignal` would have been named something like `StateSignal` and `ReadWriteSignal` -> `MutableStateSignal` then I would probably not have been advocating for this as much. _"A foolish consistency is the hobgoblin of little minds", though._

**Comment by mansbernhardt** [comment]
@CarlEkman If we go we a `WriteSignal`, we should probably also add a `FiniteWriteSignal`. So if you consider `FiniteSignal` (a quite late addition, as before it, all signals could terminate), the symmetry we once had with the write signal (then called sink) perhaps no longer apply?

**Comment by CarlEkman** [comment]
Haha, good point @mansbernhardt! Yes by my logic I guess we should add `FiniteReadSignal` and `FiniteReadWriteSignal` as well... Just from curiosity - how come you decided to make standard signals infinite by default?

**Comment by andriichernenko** [comment]
@mansbernhardt 
> What I don't like about exposing inputSignal as a WriteSignal is that the "signal" part of it is an implementation detail that we are leaking

This made me revisit the definition of "leaking an implementation detail". I found [an article](https://enterprisecraftsmanship.com/2016/07/27/what-is-an-implementation-detail/) that explains this pretty well:

> A good rule of thumb here is to look at the number of operations the client code has to invoke on the class in order to achieve a single goal. If this number is higher than 1, then the class is most likely exposing some implementation details. Ideally, any individual goal the outside layer has to accomplish should be solved with a single operation only.

If we apply this rule to our example, the number of operations is 1 both when `inputSignal` is a function and when it is a `WriteSignal`, so I don't think we're leaking any implementation details here. Or did you have some other definition in mind?

I would argue that making `inputSignal` a function or a signal is not an implementation detail, but rather a matter of choosing an abstraction. And since the outputs of `ViewModels` are signals anyway, making the inputs signals as well results in more consistent API. Of course, as you showed above, Flow provides pretty good support for using functions as inputs (there is a `.bindTo()` overload that accepts functions, and that clever `Callbacker` trick can make derivation of output signals much easier), but I think this abstraction is just less intuitive.

-----

As for the `internalSignal`, this is just an intermediate state derived from other input signals but not exposed as an output. It can actually be stored locally in the initializer:

```swift
struct ViewModel {
    let inputSignal: WriteSignal<...>
    let outputSignal: ReadSignal<...>
}

extension ViewModel {
    init() {
        let internalSignal: ReadSignal<...> = // ... (some combination of other input signals)

        outputSignal = inputSignal
            .throttle(0.5)
            .withLatestFrom(internalSignal)
            .map { (_, value2) in ... }
    }
}
```
We can probably disregard it.

**Comment by mansbernhardt** [comment]
@CarlEkman in my experience it's really uncommon with signals that terminates. This is especially true in Flow where we use futures instead of signals to model asynchronous operations, and it's typically these that might fail. However, it is useful to be able to model terminating signals as well and also make it explicit in the type system when applying a transform such as `take()` that it will become finite.

Hence similar as Flow makes a difference between a `ReadSignal` and `Signal` to better state for the reader what to expect, `FiniteSignal` fulfils a similar purpose. As been discussed in the thread so far, I think it's useful be as explicit about your types/APIs as possible as it make it easier to reason about them.

**Comment by mansbernhardt** [comment]
@andriichernenko perhaps I'm using incorrect wording here. What I'm trying to say is that we should be as explicit about our types as we can. We don't use optional everywhere. Instead we try to avoid them when we know a value can't be nil. Similarly if you only expose a function to report your changes, there is not much else you can do, where as using a `WriteSignal` makes you wonder why it's not just a function, perhaps I'm supposed to listen on changes introduced by some one else? But the reason for using a `WriteSignal` in your example is to is to make it more `convenient` to your current implementation, hence you are leaking implementation details about your current implementation. If you later on changes your implementation will you change the exposed type to so something else?

**Comment by jdmoreira** [comment]
In RxSwift you can take a write signal - `PublishSubject`  and call `.asObserver` on it and you would get an any `AnyObserver` that is a sink only. No `Signal`/`Observable` component to it. @mansbernhardt in that case the Signal abstraction doesn't leak because you can't listen to the values of an `AnyObserver`. It's input only.

Let's imagine we have a protocol named `Sink` where we can only write to - never read or listen to changes.

Taking @andriichernenko  example and modify it a bit
```swift
struct ViewModel {
    private let _inputSignal: ReadWriteSignal<...>
    var inputSignal: Sink<...> {
        _inputSignal.asSink
    }
    let outputSignal: ReadSignal<...>
}

extension ViewModel {
    init() {
        let internalSignal: ReadSignal<...> = // ... (some combination of other input signals)

        outputSignal = inputSignal
            .throttle(0.5)
            .withLatestFrom(internalSignal)
            .map { (_, value2) in ... }
    }
}
```

`Signal` doesn't leak anymore. Of course the case could be made that we could achieve the exact same thing with a function and a `Callbacker`.

**Comment by andriichernenko** [comment]
@mansbernhardt Oh, now I get what you mean, and it's a fair point, the clients are not supposed to be able to subscribe to input signals of a view model. As @jdmoreira pointed out, this could be solved by a protocol exposing only the writing capabilities. However, introducing such protocol feels like a much more serious change that might prompt redesign of the whole Flow API. I guess I should try using the combination of functions and `Callbacker`s before I understand if we need this change or not.
